diff --git a/.flaskenv b/.flaskenv
new file mode 100644
index 00000000..6006c26d
--- /dev/null
+++ b/.flaskenv
@@ -0,0 +1 @@
+FLASK_APP=microblog.py
diff --git a/app/__init__.py b/app/__init__.py
new file mode 100644
index 00000000..96c8ef5d
--- /dev/null
+++ b/app/__init__.py
@@ -0,0 +1,5 @@
+from flask import Flask
+
+app = Flask(__name__)
+
+from app import routes
diff --git a/app/routes.py b/app/routes.py
new file mode 100644
index 00000000..69701e32
--- /dev/null
+++ b/app/routes.py
@@ -0,0 +1,7 @@
+from app import app
+
+
+@app.route('/')
+@app.route('/index')
+def index():
+    return "Hello, World!"
diff --git a/microblog.py b/microblog.py
new file mode 100644
index 00000000..d099b92d
--- /dev/null
+++ b/microblog.py
@@ -0,0 +1 @@
+from app import app
